### PR TITLE
8333884: MemorySegment::reinterpret removes read-only property

### PR DESCRIPTION
**Pull request by minborg**
This PR proposes to retain the read-only state when any of the `MemorySegment::reinterpret` methods is called.

Previously, the read-only state was lost and the returned `MemorySegment` was always writable regardless of the original segment's read-only state.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change requires CSR request [JDK-8333892](https://bugs.openjdk.org/browse/JDK-8333892) to be approved
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issues
 * [JDK-8333884](https://bugs.openjdk.org/browse/JDK-8333884): MemorySegment::reinterpret removes read-only property (**Bug** - P3)(⚠️ The fixVersion in this issue is [23, 24] but the fixVersion in .jcheck/conf is 24, a new backport will be created when this pr is integrated.)
 * [JDK-8333892](https://bugs.openjdk.org/browse/JDK-8333892): MemorySegment::reinterpret removes read-only property (**CSR**)

### Reviewers
 * [Maurizio Cimadamore](https://openjdk.org/census#mcimadamore) (@mcimadamore - **Reviewer**) 🔄 Re-review required (review applies to [538eb2ed](https://git.openjdk.org/jdk/pull/19629/files/538eb2ed887785f1ed6bc3fd2be21a759f4b6cca))
 * [Jorn Vernee](https://openjdk.org/census#jvernee) (@JornVernee - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/19629/head:pull/19629` \
`$ git checkout pull/19629`

Update a local copy of the PR: \
`$ git checkout pull/19629` \
`$ git pull https://git.openjdk.org/jdk.git pull/19629/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 19629`

View PR using the GUI difftool: \
`$ git pr show -t 19629`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/19629.diff">https://git.openjdk.org/jdk/pull/19629.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/19629#issuecomment-2158376099)

### CODE REVIEW

File: src/java.base/share/classes/java/lang/foreign/MemorySegment.java [hunk 1]
```diff
@@ -704,8 +704,8 @@ public sealed interface MemorySegment permits AbstractMemorySegmentImpl {
     MemorySegment asSlice(long offset);
 
     /**
-     * Returns a new memory segment that has the same address and scope as this segment,
-     * but with the provided size.
+     * Returns a new memory segment that has the same address, scope, and
+     * read-only state as this segment, but with the provided size.
```
Comment (mcimadamore) [inline src/java.base/share/classes/java/lang/foreign/MemorySegment.java:708]: In other parts of the javadoc we say something like this:

```
* If the buffer is {@linkplain Buffer#isReadOnly() read-only}, the resulting segment
     * is also {@linkplain ByteBuffer#isReadOnly() read-only}
```

I think this is better, as read-only is a bit of info that is not IMHO worth spelling out in the method javadoc first para,

Also, the term "read-only state" does not appear anywhere else in the FFM javadoc, so I'd recommend against using it. We can just say "if this segment is (link) read-only, the returned segment is read-only", or something like that.

**Review by mcimadamore — COMMENTED** [review]
Note that `asSlice` methods also lacks a similar guarantee on read-only (but the impl works fine there)

File: src/java.base/share/classes/java/lang/foreign/MemorySegment.java [hunk 2]
```diff
@@ -706,6 +706,9 @@ public sealed interface MemorySegment permits AbstractMemorySegmentImpl {
     /**
      * Returns a new memory segment that has the same address and scope as this segment,
      * but with the provided size.
+     * <p>
+     * If, and only if, this segment is {@linkplain MemorySegment#isReadOnly() read-only},
```
Comment (mcimadamore) [inline src/java.base/share/classes/java/lang/foreign/MemorySegment.java:710]: Correct, but the `, and only if, ` seems a tad too pedantic?

**Review by mcimadamore — APPROVED** [review]

**Comment by minborg** [comment]
> Note that `asSlice` methods also lacks a similar guarantee on read-only (but the impl works fine there)

Even though not mentioned in the original issue, we could add documentation for this here as well.

**Comment by mcimadamore** [comment]
> > Note that `asSlice` methods also lacks a similar guarantee on read-only (but the impl works fine there)
> 
> Even though not mentioned in the original issue, we could add documentation for this here as well. Or maybe via https://bugs.openjdk.org/browse/JDK-8333886 ?

Or a new issue - I'm fine with both approaches. Whatever requires less work. Since you are already adding these "if this segment is read-only...  ", maybe it would be simpler to review as part of this PR?

**Review by JornVernee — APPROVED** [review]

**Comment by JornVernee** [comment]
Thanks for fixing!

**Review by JornVernee — APPROVED** [review]

**Pull request closed without merging**